<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="scregclust">
<title>Demonstration of workflow • scregclust</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Demonstration of workflow">
<meta property="og:description" content="scregclust">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">scregclust</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.8</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-vignettes">Vignettes</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-vignettes">
    <a class="dropdown-item" href="../articles/pbmc.html">Demonstration of workflow</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/scmethods/scregclust" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Demonstration of workflow</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/scmethods/scregclust/blob/HEAD/vignettes/pbmc.Rmd" class="external-link"><code>vignettes/pbmc.Rmd</code></a></small>
      <div class="d-none name"><code>pbmc.Rmd</code></div>
    </div>

    
    
<p>The methods below are described in our article</p>
<blockquote>
<p>Larsson I &amp; Held F, et al. (2023) Reconstructing the regulatory
programs underlying the phenotypic plasticity of neural cancers.
Preprint available at <a href="https://www.biorxiv.org/content/10.1101/2023.03.10.532041v1" class="external-link">bioRxiv</a>;
2023.03.10.532041.</p>
</blockquote>
<p>Here we demonstrate the scregclust workflow using the PBMC data from
10X Genomics (available <a href="https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-3-k-1-standard-2-0-0" class="external-link">here</a>).
This is the same data used in an <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial" class="external-link">introductory
vignette</a> for the Seurat package. We use <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a> for pre-processing of
the data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scmethods.github.io/scregclust/">scregclust</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="download-the-data">Download the data<a class="anchor" aria-label="anchor" href="#download-the-data"></a>
</h2>
<p>We are focusing here on the filtered feature barcode matrix available
as an HDF5 file from the website linked above. The data can be
downloaded manually or using R.</p>
<p>However you obtain the data, the code below assumes that the HDF5
file containing it is placed in the same folder as this script with the
name
<code>pbmc_granulocyte_sorted_3k_filtered_feature_bc_matrix.h5</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://cf.10xgenomics.com/samples/cell-arc/2.0.0/"</span>,</span>
<span>  <span class="st">"pbmc_granulocyte_sorted_3k/"</span>,</span>
<span>  <span class="st">"pbmc_granulocyte_sorted_3k_filtered_feature_bc_matrix.h5"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="st">"pbmc_granulocyte_sorted_3k_filtered_feature_bc_matrix.h5"</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">path</span>, cacheOK <span class="op">=</span> <span class="cn">FALSE</span>, mode <span class="op">=</span> <span class="st">"wb"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-the-data-in-seurat-and-preprocess">Load the data in Seurat and preprocess<a class="anchor" aria-label="anchor" href="#load-the-data-in-seurat-and-preprocess"></a>
</h2>
<p>To perform preprocessing use Seurat to load the data. The file ships
with two modalities, “Gene Expression” and “Peaks”. We only use the
former.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span></span>
<span>  <span class="st">"pbmc_granulocyte_sorted_3k_filtered_feature_bc_matrix.h5"</span>,</span>
<span>  use.names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  unique.features <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span><span class="op">[[</span><span class="st">"Gene Expression"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; Genome matrix has multiple modalities, returning a list of matrices for this genome</span></span></code></pre></div>
<p>We create a Seurat object and follow the Seurat vignette to subset
the cells and features (genes).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>  counts <span class="op">=</span> <span class="va">pbmc_data</span>, min.cells <span class="op">=</span> <span class="fl">3</span>, min.features <span class="op">=</span> <span class="fl">200</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc</span>, pattern <span class="op">=</span> <span class="st">"^MT."</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc</span>, subset <span class="op">=</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">30</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">6000</span><span class="op">)</span></span></code></pre></div>
<p><a href="https://satijalab.org/seurat/articles/sctransform_vignette" class="external-link">SCTransform</a>
is used for variance stabilization of the data and Pearson residuals for
the 6000 most variable genes are extracted as matrix <code>z</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">pbmc</span>, variable.features.n <span class="op">=</span> <span class="fl">6000</span><span class="op">)</span></span>
<span><span class="co">#&gt; Running SCTransform on assay: RNA</span></span>
<span><span class="co">#&gt; Running SCTransform on layer: counts</span></span>
<span><span class="co">#&gt; vst.flavor='v2' set. Using model with fixed slope and excluding poisson genes.</span></span>
<span><span class="co">#&gt; Variance stabilizing transformation of count matrix of size 19168 by 2686</span></span>
<span><span class="co">#&gt; Model formula is y ~ log_umi</span></span>
<span><span class="co">#&gt; Get Negative Binomial regression parameters per gene</span></span>
<span><span class="co">#&gt; Using 2000 genes, 2686 cells</span></span>
<span><span class="co">#&gt; Found 6 outliers - those will be ignored in fitting/regularization step</span></span>
<span><span class="co">#&gt; Second step: Get residuals using fitted parameters for 19168 genes</span></span>
<span><span class="co">#&gt; Computing corrected count matrix for 19168 genes</span></span>
<span><span class="co">#&gt; Calculating gene attributes</span></span>
<span><span class="co">#&gt; Wall clock passed: Time difference of 13.02882 secs</span></span>
<span><span class="co">#&gt; Determine variable features</span></span>
<span><span class="co">#&gt; Centering data matrix</span></span>
<span><span class="co">#&gt; Set default assay to SCT</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc</span>, layer <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6000 2686</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="use-scregclust-for-clustering">Use scregclust for clustering<a class="anchor" aria-label="anchor" href="#use-scregclust-for-clustering"></a>
</h2>
<p>We then use <code>scregclust_format</code> which extracts gene
symbols from the expression matrix and determines which genes are
considered regulators. By default, transcription factors are used as
regulators. Setting <code>mode</code> to <code>"kinase"</code> uses
kinases instead of transcription factors. A list of the regulators used
internally is returned by <code><a href="../reference/get_regulator_list.html">get_regulator_list()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scregclust_format.html">scregclust_format</a></span><span class="op">(</span><span class="va">z</span>, mode <span class="op">=</span> <span class="st">"TF"</span><span class="op">)</span></span></code></pre></div>
<p><code>sample_assignment</code> is initialized to be a vector of
<code>1</code>s of length <code>ncol(z)</code> and can be filled with a
known sample grouping. Here, we do not use it and just keep it uniform
across all cells.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genesymbols</span> <span class="op">&lt;-</span> <span class="va">out</span><span class="op">$</span><span class="va">genesymbols</span></span>
<span><span class="va">sample_assignment</span> <span class="op">&lt;-</span> <span class="va">out</span><span class="op">$</span><span class="va">sample_assignment</span></span>
<span><span class="va">is_regulator</span> <span class="op">&lt;-</span> <span class="va">out</span><span class="op">$</span><span class="va">is_regulator</span></span></code></pre></div>
<p>Run <code>scregclust</code> with number of initial clusters set to 10
and test several penalties. <code>noise_threshold</code> controls the
minimum <span class="math inline">\(R^2\)</span> a gene has to achieve
across clusters. Otherwise it is placed in the noise/rag-bag cluster.
The run can be reproduced with the command below. A pre-fitted model can
be downloaded from <a href="https://github.com/sven-nelander/scregclust/raw/main/datasets/pbmc_scregclust.rds" class="external-link">GitHub</a>
for convenience.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set.seed(8374)</span></span>
<span><span class="co"># fit &lt;- scregclust(</span></span>
<span><span class="co">#   z, genesymbols, is_regulator, penalization = seq(0.1, 0.5, 0.1),</span></span>
<span><span class="co">#   n_cl = 10L, n_cycles = 50L, noise_threshold = 0.05, center = FALSE,</span></span>
<span><span class="co">#   sample_assignment = sample_assignment</span></span>
<span><span class="co"># )</span></span>
<span></span>
<span><span class="va">url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>  <span class="st">"https://github.com/sven-nelander/scregclust/raw/main/datasets/"</span>,</span>
<span>  <span class="st">"pbmc_scregclust.rds"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">path</span> <span class="op">&lt;-</span> <span class="st">"pbmc_scregclust.rds"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="va">path</span><span class="op">)</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"pbmc_scregclust.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="analysis-of-results">Analysis of results<a class="anchor" aria-label="anchor" href="#analysis-of-results"></a>
</h2>
<p>Results can be visualized easily using built-in functions. Metrics
for helping in choosing an optimal penalty can be plotted by calling
<code>plot</code> on the object returned from
<code>scregclust</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc_files/figure-html/viz-metrics-1.png" width="700"></p>
<p>The results for each penalization parameter are placed in a list,
<code>results</code>, attached to the <code>fit</code> object. So
<code>fit$results[[1]]</code> contains the results of running
<code>scregclust</code> with <code>penalization = 0.1</code>. For each
penalization parameter, the algorithm might end up finding multiple
optimal clusterings. The results for each such clustering are contained
in the list <code>output</code>. This means that
<code>fit$results[[1]]$output[[1]]</code> contains the results for the
first final clustering. More than one may be available.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">results</span>, <span class="kw">function</span><span class="op">(</span><span class="va">r</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">r</span><span class="op">$</span><span class="va">output</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2 1 1 1 1</span></span></code></pre></div>
<p>In this example, two final configurations were found for the lowest
penalization and only a single configuration for all other penalization
parameters.</p>
<p>To plot the regulator network of the first clustering for
<code>penalization = 0.1</code> the function
<code>plot_regulator_network</code> can be used.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_regulator_network.html">plot_regulator_network</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">results</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">output</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="pbmc_files/figure-html/viz-reg-network-1.png" width="700"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Felix Held, Ida Larsson, Sven Nelander.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
